src/python/pants/backend/codegen/thrift/lib/apache_thrift_gen_base.py

```diff
@@ -15,12 +15,17 @@
 from pants.option.custom_types import target_option
 from pants.task.simple_codegen_task import SimpleCodegenTask
 from pants.util.memo import memoized_property
+from pants.util.meta import classproperty
 
 
 class ApacheThriftGenBase(SimpleCodegenTask):
   # The name of the thrift generator to use. Subclasses must set.
   # E.g., java, py (see `thrift -help` for all available generators).
   thrift_generator = None
```

Comment:
Rather than defining the `gen_directory` property, you could probably replace:
```
thrift_generator = None
```
with
```
@classproperty
@abstractmethod
def thrift_generator(cls):
  """<docstring>"""
```

That would help folks extending this class to do the right thing and definitely override this variable.

Reply:
i don't understand what you mean here, for most of the thrift generators the `gen_dir` is the same as the `thrift_generator` property but in case of nodejs they are different which is why i set the default `gen_directory` class property to just return thrift generator and simply override them both in the `node_thrift_library` class.

Reply:
I believe Stu's comment doesn't change what you're trying to do. Essentially vanilla Python doesn't have a strict access modifier and this is mainly to help "folks extending this class to do the right thing"